staging/src/k8s.io/kube-aggregator/pkg/controllers/openapi/aggregator/aggregator.go

```diff
@@ -17,330 +17,248 @@ limitations under the License.
-	if existedBefore && origSpecInfo != nil && origSpecInfo.etag == specInfo.etag {
-		return nil
-	}
-	klog.V(2).Infof("Updating OpenAPI spec because %s is updated", specInfo.apiService.Name)
-	defer func(start time.Time) {
-		duration := time.Since(start)
-		klog.V(2).Infof("Finished OpenAPI spec generation after %v", duration)
-
-		reason := "add"
-		if updated {
-			reason = "update"
-		}
-
-		regenerationCounter.With(map[string]string{"apiservice": specInfo.apiService.Name, "reason": reason})
-		regenerationDurationGauge.With(map[string]string{"reason": reason}).Set(duration.Seconds())
-	}(time.Now())
-	if err := s.updateOpenAPISpec(); err != nil {
-		if existedBefore {
-			s.openAPISpecs[specInfo.apiService.Name] = origSpecInfo
-		} else {
-			delete(s.openAPISpecs, specInfo.apiService.Name)
+	return cached.NewListMerger(func(results []cached.Result[*spec.Swagger]) cached.Result[*spec.Swagger] {
+		var merged *spec.Swagger
+		etags := make([]string, 0, len(results))
+		for _, specInfo := range results {
+			result := specInfo.Get()
+			if result.Err != nil {
+				// APIService name and err message will be included in
+				// the error message as part of newNameDecoratedErrorCache
+				klog.Warning(result.Err)
+				continue
+			}
+			if merged == nil {
+				merged = &spec.Swagger{}
+				*merged = *result.Data
+				merged.Paths = nil
+				merged.Definitions = nil
+			}
+			etags = append(etags, result.Etag)
+			if err := aggregator.MergeSpecsIgnorePathConflict(merged, result.Data); err != nil {
+				return cached.NewResultErr[*spec.Swagger](fmt.Errorf("failed to build merge specs: %v", err))
+			}
 		}
-		return err
-	}
-	return nil
+		// Printing the etags map is stable.
+		return cached.NewResultOK(merged, fmt.Sprintf("%x", sha256.Sum256([]byte(fmt.Sprintf("%#v", etags)))))
+	}, caches)
 }
 
-// tryDeleteServiceSpecs tries delete specified specInfo from openAPISpecs map, and keeps the map intact
-// if the update fails.
-func (s *specAggregator) tryDeleteServiceSpecs(apiServiceName string) error {
-	orgSpecInfo, exists := s.openAPISpecs[apiServiceName]
+// updateServiceLocked updates the spec cache by downloading the latest
+// version of the spec.
+func (s *specAggregator) updateServiceLocked(name string) error {
+	specInfo, exists := s.openAPISpecs[name]
 	if !exists {
-		return nil
+		return ErrNotFound
 	}
-	delete(s.openAPISpecs, apiServiceName)
-	klog.V(2).Infof("Updating OpenAPI spec because %s is removed", apiServiceName)
-	defer func(start time.Time) {
-		duration := time.Since(start)
-		klog.V(2).Infof("Finished OpenAPI spec generation after %v", duration)
-
-		regenerationCounter.With(map[string]string{"apiservice": apiServiceName, "reason": "delete"})
-		regenerationDurationGauge.With(map[string]string{"reason": "delete"}).Set(duration.Seconds())
-	}(time.Now())
-	if err := s.updateOpenAPISpec(); err != nil {
-		s.openAPISpecs[apiServiceName] = orgSpecInfo
-		return err
-	}
-	return nil
+	result := specInfo.downloader.Get()
```

Comment:
Maybe comment that this is resolved directly (non-lazily) because we want to update this every minute and not do that lazily because we don't know if aggregated apiservers will be slow or fast.